### ISSUE TITLE
S.of(context) always return null if on MaterialApp.title or if localeResolutionCallback is set

### ISSUE THREAD

**Issue by dpedrinha** [flutter/flutter#39593]
I found two bugs on localization.

I just created an app, added the app_title string to the arb file and it's crashing. Seems like MaterialApp.title is calling the localization before it's ready.

`class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: S.of(context).app_name,
      localizationsDelegates: [S.delegate],
      supportedLocales: S.delegate.supportedLocales,`

With this setup, if I change the title to and empty string "" i18n works on other widgets, including the Scaffold.AppBar.title property.

But if I set: 

> localeResolutionCallback: S.delegate.resolution(fallback: Locale('en')),

Then I get the same error everywhere.

## Steps to Reproduce

1. ... Create a new app
2. ... Extract the string resource
3. ... Change the MaterialApp.title to this resource with S.of(context).app_name or set it on any other Text widget

## Logs

``` I/flutter ( 7651): Warning: This application's locale, en, is not supported by all of its
I/flutter ( 7651): localization delegates.
I/flutter ( 7651): > A S delegate that supports the en locale was not found.
I/flutter ( 7651): See https://flutter.dev/tutorials/internationalization/ for more
I/flutter ( 7651): information about configuring an app's locale, supportedLocales,
I/flutter ( 7651): and localizationsDelegates parameters.
I/flutter ( 7651): ════════
I/flutter ( 7651): ══╡ EXCEPTION CAUGHT BY WIDGETS LIBRARY ╞═══════════════════════════════════════════════════════════
I/flutter ( 7651): The following NoSuchMethodError was thrown building MainView(dirty, dependencies:
I/flutter ( 7651): [_LocalizationsScope-[GlobalKey#42a20]], state: _MainViewState#40827):
I/flutter ( 7651): The getter 'app_name' was called on null.
I/flutter ( 7651): Receiver: null
I/flutter ( 7651): Tried calling: app_name
I/flutter ( 7651): User-created ancestor of the error-causing widget was:
I/flutter ( 7651):   MaterialApp file:///D:/github-repo/aaflutter/lib/main.dart:11:12
I/flutter ( 7651): When the exception was thrown, this was the stack:
I/flutter ( 7651): #0      Object.noSuchMethod (dart:core-patch/object_patch.dart:50:5)
I/flutter ( 7651): #1      _MainViewState.build (package:aaflutter/views/view_main.dart:48:35)
I/flutter ( 7651): #2      StatefulElement.build (package:flutter/src/widgets/framework.dart:4039:27)
I/flutter ( 7651): #3      ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:3933:15)
I/flutter ( 7651): #4      Element.rebuild (package:flutter/src/widgets/framework.dart:3730:5)
I/flutter ( 7651): #5      StatefulElement.update (package:flutter/src/widgets/framework.dart:4112:5)
I/flutter ( 7651): #6      Element.updateChild (package:flutter/src/widgets/framework.dart:2885:15)
I/flutter ( 7651): #7      SingleChildRenderObjectElement.update (package:flutter/src/widgets/framework.dart:5126:14)
I/flutter ( 7651): #8      Element.updateChild (package:flutter/src/widgets/framework.dart:2885:15)
I/flutter ( 7651): #9      ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:3953:16)
I/flutter ( 7651): #10     Element.rebuild (package:flutter/src/widgets/framework.dart:3730:5)
I/flutter ( 7651): #11     StatelessElement.update (package:flutter/src/widgets/framework.dart:4008:5)
I/flutter ( 7651): #12     Element.updateChild (package:flutter/src/widgets/framework.dart:2885:15)
I/flutter ( 7651): #13     SingleChildRenderObjectElement.update (package:flutter/src/widgets/framework.dart:5126:14)
I/flutter ( 7651): #14     Element.updateChild (package:flutter/src/widgets/framework.dart:2885:15)
I/flutter ( 7651): #15     SingleChildRenderObjectElement.update (package:flutter/src/widgets/framework.dart:5126:14)
I/flutter ( 7651): #16     Element.updateChild (package:flutter/src/widgets/framework.dart:2885:15)
I/flutter ( 7651): #17     SingleChildRenderObjectElement.update (package:flutter/src/widgets/framework.dart:5126:14)
I/flutter ( 7651): #18     Element.updateChild (package:flutter/src/widgets/framework.dart:2885:15)
I/flutter ( 7651): #19     SingleChildRenderObjectElement.update (package:flutter/src/widgets/framework.dart:5126:14)
I/flutter ( 7651): #20     Element.updateChild (package:flutter/src/widgets/framework.dart:2885:15)
I/flutter ( 7651): #21     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:3953:16)
I/flutter ( 7651): #22     Element.rebuild (package:flutter/src/widgets/framework.dart:3730:5)
I/flutter ( 7651): #23     StatefulElement.update (package:flutter/src/widgets/framework.dart:4112:5)
I/flutter ( 7651): #24     Element.updateChild (package:flutter/src/widgets/framework.dart:2885:15)
I/flutter ( 7651): #25     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:3953:16)
I/flutter ( 7651): #26     Element.rebuild (package:flutter/src/widgets/framework.dart:3730:5)
I/flutter ( 7651): #27     StatelessElement.update (package:flutter/src/widgets/framework.dart:4008:5)
I/flutter ( 7651): #28     Element.updateChild (package:flutter/src/widgets/framework.dart:2885:15)
I/flutter ( 7651): #29     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:3953:16)
I/flutter ( 7651): #30     Element.rebuild (package:flutter/src/widgets/framework.dart:3730:5)
I/flutter ( 7651): #31     StatefulElement.update (package:flutter/src/widgets/framework.dart:4112:5)
I/flutter ( 7651): #32     Element.updateChild (package:flutter/src/widgets/framework.dart:2885:15)
I/flutter ( 7651): #33     SingleChildRenderObjectElement.update (package:flutter/src/widgets/framework.dart:5126:14)
I/flutter ( 7651): #34     Element.updateChild (package:flutter/src/widgets/framework.dart:2885:15)
I/flutter ( 7651): #35     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:3953:16)
I/flutter ( 7651): #36     Element.rebuild (package:flutter/src/widgets/framework.dart:3730:5)
I/flutter ( 7651): #37     ProxyElement.update (package:flutter/src/widgets/framework.dart:4246:5)
I/flutter ( 7651): #38     _InheritedNotifierElement.update (package:flutter/src/widgets/inherited_notifier.dart:94:11)
I/flutter ( 7651): #39     Element.updateChild (package:flutter/src/widgets/framework.dart:2885:15)
I/flutter ( 7651): #40     SingleChildRenderObjectElement.update (package:flutter/src/widgets/framework.dart:5126:14)
I/flutter ( 7651): #41     Element.updateChild (package:flutter/src/widgets/framework.dart:2885:15)
I/flutter ( 7651): #42     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:3953:16)
I/flutter ( 7651): #43     Element.rebuild (package:flutter/src/widgets/framework.dart:3730:5)
I/flutter ( 7651): #44     StatefulElement.update (package:flutter/src/widgets/framework.dart:4112:5)
I/flutter ( 7651): #45     Element.updateChild (package:flutter/src/widgets/framework.dart:2885:15)
I/flutter ( 7651): #46     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:3953:16)
I/flutter ( 7651): #47     Element.rebuild (package:flutter/src/widgets/framework.dart:3730:5)
I/flutter ( 7651): #48     StatelessElement.update (package:flutter/src/widgets/framework.dart:4008:5)
I/flutter ( 7651): #49     Element.updateChild (package:flutter/src/widgets/framework.dart:2885:15)
I/flutter ( 7651): #50     SingleChildRenderObjectElement.update (package:flutter/src/widgets/framework.dart:5126:14)
I/flutter ( 7651): #51     Element.updateChild (package:flutter/src/widgets/framework.dart:2885:15)
I/flutter ( 7651): #52     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:3953:16)
I/flutter ( 7651): #53     Element.rebuild (package:flutter/src/widgets/framework.dart:3730:5)
I/flutter ( 7651): #54     ProxyElement.update (package:flutter/src/widgets/framework.dart:4246:5)
I/flutter ( 7651): #55     Element.updateChild (package:flutter/src/widgets/framework.dart:2885:15)
I/flutter ( 7651): #56     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:3953:16)
I/flutter ( 7651): #57     Element.rebuild (package:flutter/src/widgets/framework.dart:3730:5)
I/flutter ( 7651): #58     BuildOwner.buildScope (package:flutter/src/widgets/framework.dart:2340:33)
I/flutter ( 7651): #59     _WidgetsFlutterBinding&BindingBase&GestureBinding&ServicesBinding&SchedulerBinding&PaintingBinding&SemanticsBinding&RendererBinding&WidgetsBinding.drawFrame (package:flutter/src/widgets/binding.dart:701:20)
I/flutter ( 7651): #60     _WidgetsFlutterBinding&BindingBase&GestureBinding&ServicesBinding&SchedulerBinding&PaintingBinding&SemanticsBinding&RendererBinding._handlePersistentFrameCallback (package:flutter/src/rendering/binding.dart:285:5)
I/flutter ( 7651): #61     _WidgetsFlutterBinding&BindingBase&GestureBinding&ServicesBinding&SchedulerBinding._invokeFrameCallback (package:flutter/src/scheduler/binding.dart:1033:15)
I/flutter ( 7651): #62     _WidgetsFlutterBinding&BindingBase&GestureBinding&ServicesBinding&SchedulerBinding.handleDrawFrame (package:flutter/src/scheduler/binding.dart:975:9)
I/flutter ( 7651): #63     _WidgetsFlutterBinding&BindingBase&GestureBinding&ServicesBinding&SchedulerBinding.scheduleWarmUpFrame.<anonymous closure> (package:flutter/src/scheduler/binding.dart:784:7)
I/flutter ( 7651): #65     _Timer._runTimers (dart:isolate-patch/timer_impl.dart:382:19)
I/flutter ( 7651): #66     _Timer._handleMessage (dart:isolate-patch/timer_impl.dart:416:5)
I/flutter ( 7651): #67     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:172:12)
I/flutter ( 7651): (elided one frame from package dart:async-patch)
```

```Analyzing aaflutter...
No issues found! (ran in 3.7s)

```

```D:\DEV\flutter\bin\flutter.bat doctor --verbose
[√] Flutter (Channel beta, v1.8.3, on Microsoft Windows [Version 10.0.17763.678], locale pt-BR)
    • Flutter version 1.8.3 at D:\DEV\flutter
    • Framework revision e4ebcdf6f4 (5 weeks ago), 2019-07-27 11:48:24 -0700
    • Engine revision 38ac5f30a7
    • Dart version 2.5.0 (build 2.5.0-dev.1.0 0ca1582afd)

[√] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
    • Android SDK at C:\Users\dpedr\AppData\Local\Android\sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-28, build-tools 28.0.3
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b03)
    • All Android licenses accepted.

[√] Android Studio (version 3.5)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin version 38.2.3
    • Dart plugin version 191.8423
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b03)

[!] VS Code (version 1.37.0)
    • VS Code at C:\Users\dpedr\AppData\Local\Programs\Microsoft VS Code
    X Flutter extension not installed; install from
      https://marketplace.visualstudio.com/items?itemName=Dart-Code.flutter

[√] Connected device (1 available)
    • Android SDK built for x86 • emulator-5554 • android-x86 • Android 5.1.1 (API 22) (emulator)

! Doctor found issues in 1 category.
Process finished with exit code 0
```

**Comment by HansMuller** [comment]
Thanks for the report. If you can provide small self-contained app that demonstrates the problem (https://gist.github.com/ is a useful way to share the code)  that will help us with looking into it.

**Comment by dpedrinha** [comment]
> Thanks for the report. If you can provide small self-contained app that demonstrates the problem (https://gist.github.com/ is a useful way to share the code) that will help us with looking into it.

There you go.

[https://github.com/dpedrinha/flutter_i18n_bug.git](https://github.com/dpedrinha/flutter_i18n_bug.git)

Check the TODO list. I explained everything on the comments and marked them as TODOs

**Comment by GaspardMerten** [comment]
What if you used a Builder to return your title? You would access a lower context. 

It's normal that you can not access an InheritedWidget from the context where it is declared

**Comment by Aresjr** [comment]
Had the same problem, what I did was create a class to extend "S", override the method "of", which is the one that is returning null and use the "current" attribute of "S".

```dart
class R extends S {
  static of(BuildContext context) {
    return S.current;
  }
}
```

And now I use 
```dart
R.of(context).app_name
```
Guess what? It worked!
I tested it by changing the language and opening the app again and it works like a charm, changing the language properly. Somehow the property "current" keeps the right implementation of "of", I just used it to help me.
Hope it helps.

**Comment by Aresjr** [comment]
Had the same problem, what I did was create a class to extend "S", override the method "of", which is the one that is returning null and use the "current" attribute of "S".

```dart
class R extends S {
  static of(BuildContext context) {
    return S.current;
  }
}
```

And now I use

```dart
R.of(context).app_name
```

Guess what? It worked! I tested it by changing the language and opening the app again and it works like a charm, changing the language properly. Somehow the property "current" keeps the right implementation of "of", I just used it to help me. Hope it helps.
